network/src/transport.rs

```diff
@@ -207,193 +214,667 @@ pub async fn perform_handshake<T: TSocket>(
+                let listener_peer_id =
+                    identity_pubkey_to_peer_id(None, &listener_key.public_key()).unwrap();
+                let dialer_peer_id =
+                    identity_pubkey_to_peer_id(None, &dialer_key.public_key()).unwrap();
+
+                (listener_peer_id, dialer_peer_id, None)
+            }
+        };
+
+        let supported_protocols = SupportedProtocols::from(
+            [ProtocolId::ConsensusRpc, ProtocolId::DiscoveryDirectSend].iter(),
+        );
+
+        let listener_transport = LibraNetTransport::new(
+            base_transport.clone(),
+            listener_key,
+            trusted_peers.clone(),
+            HANDSHAKE_VERSION,
+            supported_protocols.clone(),
+        );
+
+        let dialer_transport = LibraNetTransport::new(
+            base_transport,
+            dialer_key,
+            trusted_peers.clone(),
+            HANDSHAKE_VERSION,
+            supported_protocols.clone(),
+        );
+
+        (
+            rt,
+            (listener_peer_id, listener_transport),
+            (dialer_peer_id, dialer_transport),
+            trusted_peers,
+            supported_protocols,
+        )
+    }
+
+    async fn write_read_msg(socket: &mut impl TSocket, msg: &[u8]) -> Bytes {
+        write_u16frame(socket, msg).await.unwrap();
+        socket.flush().await.unwrap();
+
+        let mut buf = BytesMut::new();
+        read_u16frame(socket, &mut buf).await.unwrap();
+        buf.freeze()
+    }
+
+    fn expect_memory_noise_addr(addr: &NetworkAddress) {
+        assert!(
+            matches!(addr.as_slice(), [Memory(_), NoiseIK(_), Handshake(_)]),
+            "addr: '{}'",
+            addr
+        );
+    }
+
+    fn expect_ip4_tcp_noise_addr(addr: &NetworkAddress) {
+        assert!(
+            matches!(addr.as_slice(), [Ip4(_), Tcp(_), NoiseIK(_), Handshake(_)]),
+            "addr: '{}'",
+            addr
+        );
+    }
+
+    fn test_transport_success<TTransport>(
+        base_transport: TTransport,
+        auth: Auth,
+        listen_addr: &str,
+        expect_formatted_addr: fn(&NetworkAddress),
+    ) where
+        TTransport: Transport<Error = io::Error> + Clone,
+        TTransport::Output: TSocket,
+        TTransport::Outbound: Send + 'static,
+        TTransport::Inbound: Send + 'static,
+        TTransport::Listener: Send + 'static,
+    {
+        let (
+            mut rt,
+            (listener_peer_id, listener_transport),
+            (dialer_peer_id, dialer_transport),
+            _trusted_peers,
+            supported_protocols,
+        ) = setup(base_transport, auth);
+
+        let (mut inbounds, listener_addr) = rt.enter(|| {
+            listener_transport
+                .listen_on(listen_addr.parse().unwrap())
+                .unwrap()
+        });
+        expect_formatted_addr(&listener_addr);
+        let supported_protocols_clone = supported_protocols.clone();
+
+        let listener_task = async move {
```

Comment:
could we put the comment at the top saying exactly what's happening in here, it is a bit long so it helps me re-focus as I read the code, I suspect other people might benefit

very long functions are difficult to parse